### PR TITLE
Simple unet training

### PR DESCRIPTION
**Pull request by Lenoush**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,3 +21,4 @@ docs/generated/
 docs_build/
 
 src/mrinufft/_version.py
+mrinufft_learn_unet.gif
```
Comment (chaithyagr) [inline .gitignore:24]: Please remove this. Just delete the file.

File: examples/example_FastMRI_Unet.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
+# %%
+"""Simple UNet model."""
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:2]: Add more details. Explain the model, cite a reference to it (check fastMRI). Mention the model is pulled from fastMRI

File: examples/example_FastMRI_Unet.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        image = self.operator.adj_op(kspace)
+        recon = self.unet(image.float()).abs()
+        recon /= torch.mean(recon)
+        return recon
+
+
+# %%
+# Util function to plot the state of the model
+def plot_state(axs, mri_2D, traj, recon, loss=None, save_name=None):
+    """Graphique.
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:50]: English :P

File: src/mrinufft/operators/base.py [hunk 4]
```diff
@@ -497,6 +497,53 @@ def get_lipschitz_cst(self, max_iter=10, **kwargs):
             tmp_op = self
         return power_method(max_iter, tmp_op)
 
+    @with_numpy
```
Comment (chaithyagr) [inline src/mrinufft/operators/base.py:500]: Maybe rebase your changes to master.

**Review by chaithyagr — CHANGES_REQUESTED** [review]
This should be fairly quick.
Please make a commit with "\!docs_build", to ensure we build docs for verification

File: src/mrinufft/operators/base.py (hunk 4 above)
Comment (paquiteau) [reply src/mrinufft/operators/base.py:500]: The `with_` PR (#194) is not merged yet

File: examples/example_FastMRI_Unet.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        and Patricia Johnson and Mary Bruno and Marc Parente and Krzysztof J. Geras 
+        and Joe Katsnelson and Hersh Chandarana and Zizhao Zhang and Michal Drozdzal 
+        and Adriana Romero and Michael Rabbat and Pascal Vincent and Nafissa Yakubova 
+        and James Pinkerton and Duo Wang and Erich Owens and C. Lawrence Zitnick 
+        and Michael P. Recht and Daniel K. Sodickson and Yvonne W. Lui},
+    journal = {ArXiv e-prints},
+    archivePrefix = "arXiv",
+    eprint = {1811.08839},
+    year={2018}
+}
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:38]: Please see other examples on how to cite.

File: examples/example_FastMRI_Unet.py
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:1]: This should probably be in examples/GPU/example_fastMRI_UNet.py

File: examples/example_FastMRI_Unet.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
+# %%
+"""
+Simple UNet model.
+
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:4]: Please make sure to give the title in the right  wayso that its picked up. Please follow https://github.com/mind-inria/mri-nufft/blob/master/examples/GPU/example_learn_samples.py


https://github.com/mind-inria/mri-nufft/blob/092c795c9fc251966ce0422ee40eb4419ff4faf1/examples/GPU/example_learn_samples.py#L1-L21

Also, write your math for what is being learnt.

File: examples/example_FastMRI_Unet.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+This model is a simplified version of the U-Net architecture,
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:5]: Please cite U net here. See how to cite here:

https://github.com/mind-inria/mri-nufft/blob/092c795c9fc251966ce0422ee40eb4419ff4faf1/examples/GPU/example_learn_samples.py#L20C108-L21C4

File: examples/example_FastMRI_Unet.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+or to recover missing parts of the image.
+
+This implementation of the UNet model was pulled from the FastMRI Facebook 
+repository, which is a collaborative research project aimed at advancing 
+the field of medical imaging using machine learning techniques.
+
+References
+----------
+- [source Unet] FastMRI Facebook: https://github.com/facebookresearch/fastMRI/blob/main/fastmri/models/unet.py
+
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:25]: references at the end please. Follow:

https://github.com/mind-inria/mri-nufft/blob/092c795c9fc251966ce0422ee40eb4419ff4faf1/examples/GPU/example_learn_samples.py#L209-L212

File: examples/example_FastMRI_Unet.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+.. warning::
+    The training process is make on one image, so the training is not optimal.
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:27]: We train on a single image here. In practice, this should be done on a database like fastMRI [cite].

File: examples/example_FastMRI_Unet.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    def forward(self, kspace):
+        """Forward pass of the model."""
+        image = self.operator.adj_op(kspace)
+        recon = self.unet(image.float()).abs()
+        recon /= torch.mean(recon)
+        return recon
+
+
+# %%
+# Util function to plot the state of the model
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:74]:
```suggestion
# Utility function to plot the state of the model
```

File: examples/example_FastMRI_Unet.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+
+# %%
+# Setup Inputs (models, trajectory and image)
+init_traj = initialize_2D_spiral(64, 256).reshape(-1, 2).astype(np.float32)
+model = Model(init_traj)
+model.eval()
+
+# %%
+# The image on which we are going to train.
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:117]:
```suggestion
# Get the image on which we will train our U-Net Mode
```

File: examples/example_FastMRI_Unet.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+mri_2D = torch.Tensor(np.flipud(bwdl.get_mri(4, "T1")[80, ...]).astype(np.complex64))[
+    None
+]
+mri_2D = mri_2D / torch.mean(mri_2D)
+kspace_mri_2D = model.operator.op(mri_2D)
+
+# Before training, here is the simple reconstruction we have using a
+# density compensated adjoint.
+old_recon = model(kspace_mri_2D)
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:126]: This is not the density compensated adjoint :P 
```suggestion
dc_adjoint = model.operator.adj_op(kspace_mri_2D)
```

File: examples/example_FastMRI_Unet.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+fig, axs = plt.subplots(1, 3, figsize=(15, 5))
+plot_state(axs, mri_2D, init_traj, old_recon)
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:128]:
```suggestion
plot_state(axs, mri_2D, init_traj, dc_adjoint)
```

File: examples/example_FastMRI_Unet.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        Path(os.getcwd()).parent.parent
+        / "docs"
+        / "generated"
+        / "autoexamples"
+        / "GPU"
+        / "images"
+    )
+    shutil.copyfile("mrinufft_learn_unet.gif", final_dir / "mrinufft_learn_unet.gif")
+except FileNotFoundError:
+    pass
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:197]: Please surround this with sphinx ignore tags, it will prevent it from appearing in docs.

See here: https://github.com/mind-inria/mri-nufft/blob/092c795c9fc251966ce0422ee40eb4419ff4faf1/examples/GPU/example_learn_samples.py#L165-L190 

The corresponding example: https://mind-inria.github.io/mri-nufft/generated/autoexamples/GPU/example_learn_samples.html#sphx-glr-generated-autoexamples-gpu-example-learn-samples-py

File: examples/example_FastMRI_Unet.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+# %%
+# Trained trajectory
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:200]: Its not trained trajectory :P 
```suggestion
# Reconstruction from partially trained U-Net model
```

File: examples/example_FastMRI_Unet.py [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+model.eval()
+kspace_mri_2D = model.operator.op(mri_2D)
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:202]:
```suggestion
```

You dont need to do it again

File: examples/example_FastMRI_Unet.py [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+# density compensated adjoint.
+old_recon = model(kspace_mri_2D)
+fig, axs = plt.subplots(1, 3, figsize=(15, 5))
+plot_state(axs, mri_2D, init_traj, old_recon)
+
+
+# %%
+# Start training loop
+epoch = 100
+optimizer = torch.optim.RMSprop(model.parameters(), lr=1e-3)
```
Comment (chaithyagr) [inline examples/example_FastMRI_Unet.py:134]: Try experiementing with higher LR, I see your loss goes down rapidly, perhaps we can go faster to reach better reconstructions?

**Review by chaithyagr — CHANGES_REQUESTED** [review]
Great work!
A bunch of important comments to handle for a clean documentations.

File: examples/GPU/example_fastMRI_UNet.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
+# %%
+r"""
+==================
+Simple UNet model.
+==================
+
+This model is a simplified version of the U-Net architecture, 
+which is widely used for image segmentation tasks.  
+This is implemented in the proprietary FASTMRI package [fastmri]_.  
+Users are encouraged to contact the authors if they want to use it.
```
Comment (paquiteau) [inline examples/GPU/example_fastMRI_UNet.py:10]: I think that's not necessary here (we are not talking about sparkling)

File: examples/GPU/example_fastMRI_UNet.py [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+        tqdms.set_postfix({"loss": loss.item()})  # Update progress bar
+        losses.append(loss.item())  # Store loss value
+
+        optimizer.zero_grad()  # Zero gradients
+        loss.backward()  # Backward pass
+        optimizer.step()  # Update weights
+
+        # Generate images for gif
+        hashed = joblib.hash((i, "learn_traj", time.time()))
+        filename = "/tmp/" + f"{hashed}.png"
```
Comment (paquiteau) [inline examples/GPU/example_fastMRI_UNet.py:164]: use tmpdir (like the other examples)

**Review by paquiteau — CHANGES_REQUESTED** [review]

**Review by paquiteau — APPROVED** [review]
LGTM
